vllm/core/scheduler.py

```diff
@@ -336,76 +300,243 @@ def _schedule(self) -> SchedulerOutputs:
+        be swapped or preempted.
+        """
+        token_budget = self.scheduler_config.max_num_batched_tokens
+
+        if self.swapped:
+            # If requests are swapped, prefill cannot be scheduled because
+            # it means there's no additional available blocks.
+            prefills = SchedulerPrefillOutputs.create_empty()
+        else:
+            # First schedule as many prefilling requests as possible,
+            # then schedule decoding requests.
+            prefills = self._schedule_prefills(token_budget)
+        token_budget -= prefills.num_batched_tokens
+
+        # Don't schedule decodes if prefills are scheduled.
+        # This policy is to maximize the throughput.
+        if len(prefills.seq_groups) > 0:
+            decodes = SchedulerDecodeOutputs.create_empty()
+        else:
+            decodes = self._schedule_decodes(token_budget)
+        token_budget -= decodes.num_batched_tokens
 
-        # Each sequence in the generation phase only takes one token slot.
-        # Therefore, the number of batched tokens is equal to the number of
-        # sequences in the RUNNING state.
-        num_batched_tokens = sum(
-            seq_group.num_seqs(status=SequenceStatus.RUNNING)
-            for seq_group in self.running)
+        # If any sequence group is preempted, do not swap in any sequence group.
+        if decodes.num_preempted_seqs > 0:
+            swapped_in = SchedulerSwappedOutputs.create_empty()
+        else:
+            swapped_in = self._schedule_swapped()
+
+        num_batched_tokens = (prefills.num_batched_tokens +
+                              decodes.num_batched_tokens +
+                              swapped_in.num_batched_tokens)
 
         scheduler_outputs = SchedulerOutputs(
-            scheduled_seq_groups=[
-                ScheduledSequenceGroup(seq_group=running_group,
-                                       token_chunk_size=1)
-                for running_group in self.running
-            ],
-            prompt_run=False,
+            scheduled_seq_groups=prefills.seq_groups + decodes.seq_groups +
+            swapped_in.seq_groups,
+            num_prefill_groups=len(prefills.seq_groups),
             num_batched_tokens=num_batched_tokens,
-            blocks_to_swap_in=blocks_to_swap_in,
-            blocks_to_swap_out=blocks_to_swap_out,
-            blocks_to_copy=blocks_to_copy,
-            ignored_seq_groups=[],
+            blocks_to_swap_in=swapped_in.blocks_to_swap_in,
+            blocks_to_swap_out=decodes.blocks_to_swap_out,
+            blocks_to_copy=merge_dicts(decodes.blocks_to_copy,
+                                       swapped_in.blocks_to_copy),
+            ignored_seq_groups=prefills.ignored_seq_groups,
         )
+
+        self.running.extend([s.seq_group for s in prefills.seq_groups])
+        self.running.extend([s.seq_group for s in decodes.seq_groups])
+        self.running.extend([s.seq_group for s in swapped_in.seq_groups])
```

Comment:
can you help me understand what clears `self.running` each step?

Reply:
it is popped out when it is scheduled from each func!

Reply:
seems not https://github.com/vllm-project/vllm/blob/810c56db6311d6e6066c18416b51801458caaf1b/vllm/core/scheduler.py#L281-L287